core/src/main/java/hudson/PluginManager.java

```diff
@@ -811,9 +838,36 @@ public void doInstall(StaplerRequest req, StaplerResponse rsp) throws IOExceptio
                 if (p == null) {
                     throw new Failure("No such plugin: " + n);
                 }
-                p.deploy(dynamicLoad);
+                
+                deployJobs.add(p.deploy(dynamicLoad));
             }
         }
+        
+        // Fire a one-off thread to wait for the plugins to be deployed and then
+        // refresh the dependant plugins list.
+        new Thread() {
+            @Override
+            public void run() {
+                INSTALLING: while (true) {
+                    for (Future<UpdateCenter.UpdateCenterJob> deployJob : deployJobs) {
+                        try {
+                            Thread.sleep(500);
+                        } catch (InterruptedException e) {
+                            LOGGER.log(SEVERE, "Unexpected error while waiting for some plugins to install. Plugin Manager state may be invalid. Please restart Jenkins ASAP.", e);
+                        }
+                        if (!deployJob.isCancelled() && !deployJob.isDone()) {
+                            // One of the plugins is not installing/canceled, so
+                            // go back to sleep and try again in a while.
+                            continue INSTALLING;
+                        }
+                    }
+                    // All the plugins are installed. It's now safe to refresh.
+                    resolveDependantPlugins();
+                    break;
+                }
+            }
+        }.start();
+        
```

Comment:
More a question than a critic actually: do we have a lot of things like that in other places in Jenkins core? 

Seems to me we might want to avoid starting threads from inside an appserver, and prefer submitting work to "managed" threads. That way we wouldn't have potential /wild/ thread running when wanting to stop/restart Jenkins and so on?